File: Practices/Algo-Questions/Stars_and_Bars/stars_and_bars.py
# -*- coding: utf-8 -*-
# @Author: Tairan Gao
# @Date:   2023-11-30 10:45:37
# @Last Modified by:   Tairan Gao
# @Last Modified time: 2023-11-30 11:17:32

from typing import List


def stars_and_bars(s: str, startIndex: List[int], endIndex: List[int]) -> List[int]:
    n = len(s)
    dp = [0 for i in range(n + 1)]

    skip = True
    cur = 0
    for i in range(1, n):
        c = s[i - 1]
        if skip and c != "|":
            continue

        skip = False
        if c == "|":
            dp[i] = dp[i - 1] + cur
            cur = 0
        else:
            cur += 1
            dp[i] = dp[i - 1]

    print(dp)
    sol = []
    for i in range(len(startIndex)):
        start = startIndex[i]
        end = endIndex[i]
        sol.append(dp[end] - dp[start])

    return sol


if __name__ == "__main__":
    s = "|**|*|*"
    startIndex = [1, 1]
    endIndex = [5, 6]
    sol = stars_and_bars(s, startIndex, endIndex)
    print(sol)

File: Practices/Algo-Questions/Stars_and_Bars/stars_and_bars.cpp
/*
 * @Author: Tairan Gao
 * @Date:   2023-11-30 11:18:07
 * @Last Modified by:   Tairan Gao
 * @Last Modified time: 2023-11-30 11:44:29
 */

#include <vector>
#include <string>
#include <iostream>

auto stars_and_bars(std::string s, std::vector<int> startIndex, std::vector<int> endIndex) -> std::vector<int>
{
    bool skip = true;
    int n = s.size();
    auto dp = std::vector<int>(n + 1, 0);

    int cur = 0;

    for (int i = 1; i < n + 1; i++)
    {
        char c = s[i - 1];
        if (c != '|' && skip)
        {
            continue;
        }
        skip = false;

        if (c == '*')
        {
            dp[i] = dp[i - 1];
            cur += 1;
        }
        else // (c == '|')
        {
            dp[i] = dp[i - 1] + cur;
            cur = 0;
        }
    }
    int m = startIndex.size();
    std::vector<int> sol;
    for (int i = 0; i < m; i++)
    {
        int start = startIndex[i];
        int end = endIndex[i];
        sol.push_back(dp[end] - dp[start]);
    }
    return sol;
}

int main()
{
    auto s = "|**|*|*";

    auto startIndex = {1, 1};
    auto endIndex = {5, 6};

    auto sol = stars_and_bars(s, startIndex, endIndex);
    for (const auto &ele : sol)
    {
        std::cout << ele << "\n";
    }
    return 0;
}